README.md
# pytest_mozwebqa

pytest_mozwebqa is a plugin for [py.test](http://pytest.org/) that provides additional features needed for Mozilla's WebQA projects.

Requires:

  * py.test
  * selenium
  * requests

Continuous Integration
----------------------
[![Build Status](https://travis-ci.org/mozilla/pytest-mozwebqa.svg?branch=master)](http://travis-ci.org/mozilla/pytest-mozwebqa)

## Installation

To install pytest-mozwebqa:

```bash
pip install pytest-mozwebqa
```

## Usage

The basic usage of this plugin comes from specifying the `mozwebqa` argument
in your test methods. Doing this will launch a web browser based on your
command line options. You will then have access to the
[Selenium WebDriver](http://selenium.googlecode.com/git/docs/api/py/webdriver_remote/selenium.webdriver.remote.webdriver.html)
client that controls the browser instance:

```python
def test_example(mozwebqa):
mozwebqa.selenium.get('http://www.example.com/')
assert mozwebqa.selenium.title == 'Example Domain'
```

For the full list of command line options, run `py.test --help`.

### Specifying a base URL

The base URL is the location for the application under test. It can be
specified in a number of ways. The simplest of these is the `--baseurl`
command line option, however it can be more convenient to create a
`mozwebqa.cfg` file:

```INI
[DEFAULT]
baseurl: 'http://www.example.com'
```

Alternatively, you can reimplement the `base_url` py.test fixture. This allows
the value of the base URL to be controlled by your tests. A common use for
this is when you want to run tests against a local copy of your application
that's started by your test framework:

```python
import pytest
@pytest.fixture(scope='session')
def base_url(live_server):
    return live_server.url
```

### Accessing the base URL

If you want to access the base URL from your test, you can do so by specifying
the `base_url` argument in your test method. This can be useful if you want to
run tests that do not require a browser instance against your application:

```python
def test_status_code(base_url):
    from urllib2 import urlopen
    assert urlopen(base_url).getcode() == 200
```

### Destructive tests

In order to prevent accidentally running destructive tests, only tests marked
as nondestructive will run by default. If you want to mark a test as
nondestructive then add the appropriate marker as shown below:

```python
import pytest
@pytest.mark.nondestructive
def test_safely(mozwebqa):
    assert True
```

If you want to run destructive tests then you can specify the `--destructive`
command line option.

### Sensitive environments

If running against a sensitive (production) environment, any destructive tests
will be skipped with an appropriate error message. You can specify a regular
expression that matches your sensitive environments using the `--sensitiveurl`
command line option. By default all URLs are considered sensitive.

Setting capabilities
--------------------

It's possible to specify additional capabilities on the command line:

### Example (accept SSL certificates)

    --capability=acceptSslCerts:true


Setting Firefox preferences
---------------------------

If you're using Firefox it's possible to set custom preferences:

### Example (disable addon compatibility checking)

    --firefoxpref=extensions.checkCompatibility.nightly:false

Specifying a Firefox profile
----------------------------

If you're using Firefox it's possible to specify an existing Firefox profile to use when starting Firefox.

### Example (use the profile located at /path/to/profile_directory)

    --profilepath='/path/to/profile_directory'

Installing Firefox extensions
-----------------------------

If you're using Firefox it's possible to install extensions when starting the browser.

### Example (install the extensions located at /path/to/ext1/ext1.xpi and /path/to/ext2/ext2.xpi)

    --extension='/path/to/ext1/ext1.xpi' --extension='/path/to/ext2/ext2.xpi'

Setting Google Chrome options
-----------------------------

If you're using Google Chrome then you can set various options on the command line using a JSON string.

Valid keys are:
 * arguments: a list of command-line arguments to use when starting Google Chrome.
 * binary_location: path to the Google Chrome executable to use.

For more details on Google Chrome options see: http://code.google.com/p/chromedriver/wiki/CapabilitiesAndSwitches

### Example (set initial homepage)

    --chromeopts='{"arguments":["homepage=http://www.example.com"]}'

Installing Google Chrome extensions
-----------------------------------

If you're using Google Chrome it's possible to install extensions when starting the browser.

### Example (install the extensions located at /path/to/ext1/ext1.crx and /path/to/ext2/ext2.crx)

    --extension='/path/to/ext1/ext1.crx' --extension='/path/to/ext2/ext2.crx'

## HTML report

If the pytest-html plugin is installed then the HTML reports will include
additional information such as the failing URL, screenshot, and page source.
For Sauce Labs, a link to the job and inline video will also be included. Check
the [pytest-html documentation](https://pypi.python.org/pypi/pytest-html) for
how to install the plugin and generate HTML reports.

## Appium

To use [Appium](http://appium.io/) instead of Selenium, simply specify the
[appropriate capabilities](http://appium.io/slate/en/master/?python#appium-server-capabilities).
For example, to run web application tests against Safari on an iPhone simulator:

```shell
py.test --browsername=Safari \
--capability=platformName:iOS \
--capability=platformVersion:8.2 \
--capability="deviceName:iPhone Simulator" \
--capability=appiumVersion:1.3.7
```

## Sauce Labs integration

To run your automated tests using [Sauce Labs](https://saucelabs.com/), you
must provide a valid username and API key. This can be done either by creating
a `setup.cfg` file with a `[saucelabs]` section or by setting the
`SAUCELABS_USERNAME` and `SAUCELABS_API_KEY` environment variables.

### Configuration

Below is an example `setup.cfg` showing the configuration options:

```INI
[saucelabs]
username = username
api-key = secret
tags = tag1, tag2
privacy = public
```

The `tags` entry is an optional comma separated list of tags that can be used
to filter the jobs in the Sauce Labs dashboard.

The `privacy` entry is used to determine who you share your Sauce Labs jobs
with. Check the
[documentation](https://saucelabs.com/docs/additional-config#sharing) for the
accepted values. If not set, this defaults to
[public restricted](https://saucelabs.com/docs/additional-config#restricted).

### Running tests

To run your automated tests, simply specify `SauceLabs` as your driver:

```shell
py.test --driver=SauceLabs --browsername=Firefox --platform="Windows 8"
```

See the [supported platforms](https://docs.saucelabs.com/reference/platforms-configurator/)
to help you with your configuration. Additional capabilities can be set using
the `--capability` comand line arguments. See the
[test configuration documentation](https://docs.saucelabs.com/reference/test-configuration/)
for full details of what can be configured.

### Privacy

You can specify the job sharing level for individual tests by setting a mark on
the test method. This takes priority over the `privacy` entry in the
configuration file:

```python
import pytest
@pytest.mark.privacy('public')
def test_public(mozwebqa):
    assert True
```

You can also explicitly mark the test as private:


```python
import pytest
@pytest.mark.privacy('private')
def test_private(mozwebqa):
    assert True
```

For the full list of accepted values, check the
[Sauce Labs documentation](https://saucelabs.com/docs/additional-config#sharing).

## BrowserStack integration

To run your automated tests using
[BrowserStack](https://www.browserstack.com/), you must provide a valid
username and access key. This can be done either by creating a `setup.cfg`
file with a `[browserstack]` section or by setting the `BROWSERSTACK_USERNAME`
and `BROWSERSTACK_ACCESS_KEY` environment variables.

### Configuration

Below is an example `setup.cfg` showing the configuration options:

```ini
[browserstack]
username = username
access-key = secret
```

### Running tests

To run your automated tests, simply specify `BrowserStack` as your driver:

```shell
py.test --driver=BrowserStack --browsername=firefox --platform=WIN8
```

See the [capabilities documentation](https://www.browserstack.com/automate/capabilities)
for additional configuration that can be set using `--capability` command line
arguments.

Using a proxy server
--------------------

If you want the browser launched to use a proxy (currently only supported by Firefox and Google Chrome) you must specify the `--proxyhost` and `--proxyport` command line arguments.

### Example (proxy is running on localhost port 8080)

    --proxyhost=localhost --proxyport=8080


CHANGELOG.md
Change Log
==========

2.0
---
* Add support for BrowserStack
* Use pytest-html for HTML reports
* Remove support for Selenium RC

1.6
---
* Use requests when sending the Sauce Labs result.

1.5
---
* Support environment variables for Selenium host and port.

1.4
---
* Only check the base URL from the main process when pytest-xdist is in use.

1.3
---
* Output request headers when initial test for baseurl fails. Fixes #130

1.2
-----
* Improved handling of Firefox preferences and additional capabilities. Fixes #103 & #104
* Skip destructive tests on firefox.com domain by default
* Use a parameterised test for sensitive URLs
* Support for Selenium EventListeningWebDriver
* Fixed UnicodeDecodeError in HTML report. Fixes #114
* Update the Selenium server used in the Travis-CI tests
* Add simple headless testing support with PhantomJS
* Load Selenium defaults from environment variables
* Adding Appium support for Saucelabs
* Allow overriding the Sauce Labs sharing

1.1.1
-----
* Open links from the HTML report in a new tab/window. Fixes #100
* Remove Sauce Labs links and videos for jobs without sessions. Fixes #101

1.1
---

* Added support for setting a proxy for Firefox, Google Chrome, and Opera
* Added support for specifying a Firefox profile
* Added support for specifying Firefox extensions
* Added support for pytest-browsermob-proxy
* Added support for pytest-zap
* Added sortable HTML reports
* Provide an option to skip the URL check

1.0
---

* First stable release
* Added support for Opera
* Compatible with pytest-timeout

0.10 (beta)
-----------

* Allow additional capabilities to be specified on the command line

0.9 (beta)
----------

* Only run nondestructive tests by default
* Skip destructive tests if running against a sensitive environment

0.8 (beta)
----------

 * Added support for privacy markers
 * Capture final URL on test failures
 * Capture log on test failures if test is marked as public
 * Capture HTML and screenshot only on test failures
 * Improved the debug gathering
 * Timeout value is now expected in seconds
 * Added support for specifying Google Chrome options

0.7 (beta)
----------

 * Added support for specifying Firefox preferences - fixes issue #27

0.6 (beta)
----------

 * By default Sauce Labs jobs are public (restricted) - fixes issue #14
 * Avoid running configuration hook for all slave processes - fixes issues #22 and #23
 * Add test durations to HTML report when running with multiple processes - fixes issue #9

0.5 (beta)
----------

 * Added a build identifier for when running with continuous integration

0.4 (beta)
----------

 * Changed default HTML report to results/index.html
 * Send pass/fail status to Sauce Labs - fixes issue #21
 * Added support for custom tags for Sauce Labs jobs
 * Added support for configuration files
 * Remove test names from Sauce Labs tags - fixes issue #20

0.3 (beta)
----------

 * Added error count to custom report - fixes issue #15
 * Catch exceptions thrown when attempting to capture screenshots - fixes issue #16

0.2 (beta)
----------

 * Added custom HTML reports

0.1 (alpha)
-----------

 * Initial release

pytest_mozwebqa.py
#!/usr/bin/env python

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import py
import re
import ConfigParser

import requests

import credentials


__version__ = '0.9'

def pytest_configure(config):
    if not hasattr(config, 'slaveinput'):

        config.addinivalue_line(
            'markers', 'nondestructive: mark the test as nondestructive. ' \
            'Tests are assumed to be destructive unless this marker is ' \
            'present. This reduces the risk of running destructive tests ' \
            'accidentally.')

        if config.option.base_url:
            r = requests.get(config.option.base_url)
            assert r.status_code == 200, 'Base URL did not return status code 200. (URL: %s, Response: %s)' % (config.option.base_url, r.status_code)

        if config.option.webqa_report_path:
            from html_report import HTMLReport
            config._html = HTMLReport(config)
            config.pluginmanager.register(config._html)

        if not config.option.run_destructive:
            if config.option.markexpr:
                config.option.markexpr = 'nondestructive and (%s)' % config.option.markexpr
            else: 
                config.option.markexpr = 'nondestructive'


def pytest_unconfigure(config):
    html = getattr(config, '_html', None)
    if html:
        del config._html
        config.pluginmanager.unregister(html)


def pytest_runtest_setup(item):
    item.debug = {
        'urls': [],
        'screenshots': [],
        'html': [],
        'logs': [],
        'network_traffic': []}
    TestSetup.base_url = item.config.option.base_url

    # consider this environment sensitive if the base url or any redirection
    # history matches the regular expression
    sensitive = False
    if TestSetup.base_url:
        r = requests.get(TestSetup.base_url)
        urls = [h.url for h in r.history] + [r.url]
        matches = [re.search(item.config.option.sensitive_url, u) for u in urls]
        sensitive = any(matches)

    destructive = 'nondestructive' not in item.keywords

    if (sensitive and destructive):
        first_match = matches[next(i for i, match in enumerate(matches) if match)]

        # skip the test with an appropriate message
        py.test.skip('This test is destructive and the target URL is ' \
                     'considered a sensitive environment. If this test is ' \
                     'not destructive, add the \'nondestructive\' marker to ' \
                     'it. Sensitive URL: %s' % first_match.string)

    if item.config.option.sauce_labs_credentials_file:
        item.sauce_labs_credentials = credentials.read(item.config.option.sauce_labs_credentials_file)

    if item.config.option.credentials_file:
        TestSetup.credentials = credentials.read(item.config.option.credentials_file)

    test_id = '.'.join(split_class_and_test_names(item.nodeid))

    if 'skip_selenium' not in item.keywords:
        if hasattr(item, 'sauce_labs_credentials'):
            from sauce_labs import Client
            TestSetup.selenium_client = Client(
                test_id,
                item.config.option,
                item.keywords,
                item.sauce_labs_credentials)
        else:
            from selenium_client import Client
            TestSetup.selenium_client = Client(
                test_id,
                item.config.option)
        TestSetup.selenium_client.start()
        item.session_id = TestSetup.selenium_client.session_id
        TestSetup.selenium = TestSetup.selenium_client.selenium
        TestSetup.timeout = TestSetup.selenium_client.timeout
        TestSetup.default_implicit_wait = TestSetup.selenium_client.default_implicit_wait
    else:
        TestSetup.selenium = None


def pytest_runtest_teardown(item):
    if hasattr(TestSetup, 'selenium') and TestSetup.selenium and 'skip_selenium' not in item.keywords:
        TestSetup.selenium_client.stop()


def pytest_runtest_makereport(__multicall__, item, call):
    report = __multicall__.execute()
    if report.when == 'call':
        report.session_id = getattr(item, 'session_id', None)
        if hasattr(TestSetup, 'selenium') and TestSetup.selenium and not 'skip_selenium' in item.keywords:
            if report.skipped and 'xfail' in report.keywords or report.failed and 'xfail' not in report.keywords:
                url = TestSetup.selenium_client.url
                url and item.debug['urls'].append(url)
                screenshot = TestSetup.selenium_client.screenshot
                screenshot and item.debug['screenshots'].append(screenshot)
                html = TestSetup.selenium_client.html
                html and item.debug['html'].append(html)
                log = TestSetup.selenium_client.log
                log and item.debug['logs'].append(log)
                report.sections.append(('pytest-mozwebqa', _debug_summary(item.debug)))
            network_traffic = TestSetup.selenium_client.network_traffic
            network_traffic and item.debug['network_traffic'].append(network_traffic)
            report.debug = item.debug
            if hasattr(item, 'sauce_labs_credentials') and report.session_id:
                result = {'passed': report.passed or (report.failed and 'xfail' in report.keywords)}
                import sauce_labs
                sauce_labs.Job(report.session_id).send_result(
                    result,
                    item.sauce_labs_credentials)
    return report


def pytest_funcarg__mozwebqa(request):
    return TestSetup(request)


def pytest_addoption(parser):
    config = ConfigParser.ConfigParser(defaults={
        'baseurl': '',
        'api': 'webdriver'
    })
    config.read('mozwebqa.cfg')

    group = parser.getgroup('selenium', 'selenium')
    group._addoption('--baseurl',
                     action='store',
                     dest='base_url',
                     default=config.get('DEFAULT', 'baseurl'),
                     metavar='url',
                     help='base url for the application under test.')
    group._addoption('--api',
                     action='store',
                     default=config.get('DEFAULT', 'api'),
                     metavar='api',
                     help="version of selenium api to use. 'rc' uses selenium rc. 'webdriver' uses selenium webdriver. (default: %default)")
    group._addoption('--host',
                     action='store',
                     default='localhost',
                     metavar='str',
                     help='host that selenium server is listening on. (default: %default)')
    group._addoption('--port',
                     action='store',
                     type='int',
                     default=4444,
                     metavar='num',
                     help='port that selenium server is listening on. (default: %default)')
    group._addoption('--driver',
                     action='store',
                     dest='driver',
                     default='Remote',
                     metavar='str',
                     help='webdriver implementation. (default: %default)')
    group._addoption('--capabilities',
                     action='store',
                     dest='capabilities',
                     metavar='str',
                     help='json string of additional capabilties to set (webdriver).')
    group._addoption('--chromepath',
                     action='store',
                     dest='chrome_path',
                     metavar='path',
                     help='path to the google chrome driver executable.')
    group._addoption('--firefoxpath',
                     action='store',
                     dest='firefox_path',
                     metavar='path',
                     help='path to the target firefox binary.')
    group._addoption('--firefoxpref',
                     action='store',
                     dest='firefox_preferences',
                     metavar='str',
                     help='json string of firefox preferences to set (webdriver).')
    group._addoption('--chromeopts',
                     action='store',
                     dest='chrome_options',
                     metavar='str',
                     help='json string of google chrome options to set (webdriver).')
    group._addoption('--browser',
                     action='store',
                     dest='browser',
                     metavar='str',
                     help='target browser (standalone rc server).')
    group._addoption('--environment',
                     action='store',
                     dest='environment',
                     metavar='str',
                     help='target environment (grid rc).')
    group._addoption('--browsername',
                     action='store',
                     dest='browser_name',
                     metavar='str',
                     help='target browser name (webdriver).')
    group._addoption('--browserver',
                     action='store',
                     dest='browser_version',
                     metavar='str',
                     help='target browser version (webdriver).')
    group._addoption('--platform',
                     action='store',
                     metavar='str',
                     help='target platform (webdriver).')
    group._addoption('--timeout',
                     action='store',
                     type='int',
                     default=60,
                     metavar='num',
                     help='timeout (in seconds) for page loads, etc. (default: %default)')
    group._addoption('--capturenetwork',
                     action='store_true',
                     dest='capture_network',
                     default=False,
                     help='capture network traffic to test_method_name.json (selenium rc). (default: %default)')
    group._addoption('--build',
                     action='store',
                     dest='build',
                     metavar='str',
                     help='build identifier (for continuous integration).')

    group = parser.getgroup('safety', 'safety')
    group._addoption('--sensitiveurl',
                     action='store',
                     dest='sensitive_url',
                     default='mozilla\.(com|org)',
                     metavar='str',
                     help='regular expression for identifying sensitive urls. (default: %default)')
    group._addoption('--destructive',
                     action='store_true',
                     dest='run_destructive',
                     default=False,
                     help='include destructive tests (tests not explicitly marked as \'nondestructive\'). (default: %default)')

    group = parser.getgroup('credentials', 'credentials')
    group._addoption("--credentials",
                     action="store",
                     dest='credentials_file',
                     metavar='path',
                     help="location of yaml file containing user credentials.")
    group._addoption('--saucelabs',
                     action='store',
                     dest='sauce_labs_credentials_file',
                     metavar='path',
                     help='credendials file containing sauce labs username and api key.')

    group = parser.getgroup("terminal reporting")
    group.addoption('--webqareport',
                    action='store',
                    dest='webqa_report_path',
                    metavar='path',
                    default='results/index.html',
                    help='create mozilla webqa custom report file at given path. (default: %default)')


def split_class_and_test_names(nodeid):
    names = nodeid.split("::")
    names[0] = names[0].replace("/", '.')
    names = [x.replace(".py", "") for x in names if x != "()"]
    classnames = names[:-1]
    classname = ".".join(classnames)
    name = names[-1]
    return (classname, name)


def _debug_summary(debug):
    summary = []
    if debug['urls']:
        summary.append('Failing URL: %s' % debug['urls'][-1])
    return '\n'.join(summary)


class TestSetup:
    '''
        This class is just used for monkey patching
    '''
    def __init__(self, request):
        self.request = request
